### DISCUSSION TITLE
Trying to define specific variants for individual corePlugins

### DISCUSSION THREAD

**Discussion by ryanaltvater** [tailwindlabs/tailwindcss#9288] [Help]
In v2 of Tailwind, it looks like we were able to specifically define which variants for which utilities were rendered in the output. Combing through the documentation for v3, I'm not seeing a way to do this other than defining the variants on a global level via `safelist`. I tried the examples below (from v2 of course) and it didn't seem to do anything.

```
// tailwind.config.js
module.exports = {
  variants: {
    backgroundColor: ['responsive', 'hover'],
    borderColor: ['responsive', 'focus'],
  },
}
```

```
// tailwind.config.js
module.exports = {
  variants: {
    extend: {
      backgroundColor: ['active'],
      // ...
      borderColor: ['focus-visible', 'first'],
      // ...
      textColor: ['visited'],
    }
  },
}
```

**Comment by wongjn** [comment]
In v3, [every variant is automatically available for every utility](https://tailwindcss.com/docs/upgrade-guide#remove-variant-configuration).

**Reply by ryanaltvater** [reply]
I'm aware of that, but wanted to ask anyway because that completely eliminates the option of choosing what variants you want for specific utilities. I'm creating a CSS file filled with utilities that are always available within our CMS, to build pages on the fly. That means for my use case I'm compiling unwanted bloat into my output, rather than having the ability to choose exactly which variants I want and/or don't want. This is the second feature from v2 that no longer exists that IMO is somewhat of a step backwards by removing flexibility. There's quite a few things that are now available by default, but users can still overwrite them. Users can disable default styles with `presets: []`, users can disable all plugins and define only the ones they want with `corePlugins: []`, and so why not have the ability to disable every variant for those that would rather choose.

I think what's happening is Tailwind is moving hard in the direction of being more of a static site tool when there's a lot of other use cases where others need a utility class framework alternative to something like Bootstrap and developers are turning to and hoping Tailwind can do both. In v2, Tailwind did tend to lean a little bit of both ways but with v3 I'm sensing less flexibility than before. I totally understand why variants would be available by default, but I think it would still be nice to have the option of disabling that and configuring which utilities get what variant(s) or not.

Hey, thanks for the reply though. Cheers! 👍

**Reply by wongjn** [reply]
You can use safelisting that can add only the variants you want per class when using [regular expressions](https://tailwindcss.com/docs/content-configuration#using-regular-expressions):
```js
module.exports = {
  safelist: [
    {
      pattern: /^bg-(red|green|blue)-100$/,
      variants: ['hover', 'focus'],
    },
  ],
  // …
};
```
This would only generate the `hover:` and `focus:` variants for the `bg-red-100`, `bg-green-100` and `bg-blue-100` classes.

**Reply by ryanaltvater** [reply]
I also tried this, but gave me an undesirable output. Here's my config, which currently gives me what I'm trying to do (minus variants)...

```
module.exports = {
    content: [],
    corePlugins: [
        'backgroundColor',
        'fontFamily',
        'fontSize',
        'fontStyle',
        'fontWeight',
        'margin',
        'padding',
        'preflight',
        'textColor',
        'zIndex',
    ],
    prefix: 'dx-',
    presets: [],
    safelist: [{
        pattern: /.*/,
        variants: [
            'sm',
            'md',
            'lg',
            'xl',
        ]
    }],
    theme: {
        ...
    },
    variantOrder: [
        'first',
        'last',
        'odd',
        'even',
        'visited',
        'checked',
        'empty',
        'read-only',
        'group-hover',
        'group-focus',
        'focus-within',
        'hover',
        'focus',
        'focus-visible',
        'active',
        'visited',
        'disabled',
    ],
}
```

I'm able to turn on everything via the `pattern`, but only include the `corePlugins` that I want. I don't have `content` to monitor because I don't want classes getting purged via PostCSS. Again, these are utilities that we want available in our CMS. The problem with adding `variants` is that given the RegEx pattern for "everything", the variants would then also be safelisted for everything. If I define very specific patterns instead like you've suggested, then my safelist is going to become very very lengthy as I'll need to safelist every class/pattern that I want to remain compiled in the output to achieve what I'm already doing. Once I create that lengthy safelist of patterns, then variants will be associated with some patterns that I want them to be associated with and some that I don't.

Based on your variant example, paired with my wildcard pattern, I get undesired hover utilities for font sizes, padding, and margin, etc. I may only want hover for background color, text color, etc.

**Reply by adamwathan** [reply]
I would just literally pass every class you want to ensure is generated into your safelist as a static string. Write a little function that loops over the utilities you need with the variants you need and outputs them as strings in that array.
